### PR TITLE
Week 4 - Lab

### PR DESCRIPTION
**Pull request by wwengm**
**Linear Algebra and Convolutions**

We start with a brief review of linear algebra and then extend the topic to convolutions using audio data as an example. key concepts like locality, stationarity and Toeplitz matrix are discussed. Then we give a live demo of convolution performance in pitch analysis. Finally, there is a short digression about the dimensionality of different data.

### CODE REVIEW

File: docs/chapters/04-1.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+title: Linear Algebra and Convolutions
+authors: Yuchi Ge, Anshan He, Shuting Gu, and Weiyang Wen
+---
+
+## [Linear Algebra Review](0:00:00-0:21:26)
+
+This part is a recap of basic linear algebra in the context of neural networks. We start with a simple hidden layer $\vec{h}$:
+$$
+\vec{h} = f(\vec{z})
+$$
```
Comment (Atcold) [inline docs/chapters/04-1.md]: You need to add an empty line before and after a centered equation, or it won't be rendered at all.
Here and all over the place.

File: docs/chapters/04-1.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
+---
+title: Linear Algebra and Convolutions
+authors: Yuchi Ge, Anshan He, Shuting Gu, and Weiyang Wen
+---
+
+## [Linear Algebra Review](0:00:00-0:21:26)
+
+This part is a recap of basic linear algebra in the context of neural networks. We start with a simple hidden layer $\vec{h}$:
```
Comment (Atcold) [inline docs/chapters/04-1.md]: Please, use `\vect{}` and add the macro to `https://github.com/Atcold/pytorch-Deep-Learning/blob/master/docs/_layouts/withmacros.html` such that it renders it as `\boldsymbol{}`.

File: docs/chapters/04-1.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+\vdots & \vdots & \ddots & \vdots \\
+a_{m1} & a_{m2} & \cdots & a_{mn} \end{matrix} \right] \left[\begin{matrix}
+x_1 \\ x_2\\\vdots \\x_n \end{matrix} \right]
+=\left[\begin{matrix}
+a^{(1)}\vec{x} \\a^{(2)}\vec{x}\\\vdots \\
+a^{(m)}\vec{x}\end{matrix} \right] =\vec{z}
+\end {aligned}
+$$
+
+To understand the meaning of this transformation, let us analyze one component of $\vec{z}$ such as $a^{(1)}\vec{x}$. For simplicity, assume $n=2$ thus $\vec{a} = (a_1,a_2)$ and $\vec{x}  = (x_1,x_2)$. We can draw $\vec{a}$ and $\vec{x}$ as vectors in 2D coordinate axis. Denote the angle between $a$ and $(0,1)$ as $\alpha$ and angle between $x$ and $(0,1)$ as $\xi$. Then with trigonometric formulas $a^{T}\vec{x}$ can be expanded as:
```
Comment (Atcold) [inline docs/chapters/04-1.md]: Transposition is achieved with `\^top`.

File: docs/chapters/04-1.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+\begin{matrix}
+a_{1n} \\
+a_{2n}\\
+\vdots \\
+a_{mn}
+\end{matrix} 
+\right]x_n
+$$
+The output is the weighted sum of the columns of the $A$ matrix. So the signal is nothing but a composition of the input.
+
```
Comment (Atcold) [inline docs/chapters/04-1.md]: One more white line.

File: docs/chapters/04-1.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
+---
+title: Linear Algebra and Convolutions
+authors: Yuchi Ge, Anshan He, Shuting Gu, and Weiyang Wen
+---
+
```
Comment (Atcold) [inline docs/chapters/04-1.md]: One more empty line.

File: docs/chapters/04-1.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+\end{bmatrix} = \begin{bmatrix}
+y_1\\
+y_2\\
+y_3\\
+y_4
+\end{bmatrix}
+$$
+In this example, the weight matrix has a size $4 \times 3$, the input vector has a size $3 \times 1$ and the output vector has a size $4 \times 1$.
+
+However, when we have audio data, the data is not only 3-sample long. Instead, the number of samples that the audio data has is much longer, which is equal to the duration of the audio (e.g. 3 seconds) times the sampling rate (e.g. 22.05 kHz). Like what is shown below, the input vector $\mathbf{x}$ will be long. Correspondingly, the weight matrix will become "fat".
```
Comment (Atcold) [inline docs/chapters/04-1.md]: `\vect{}` or `\boldsymbol{}`.

File: docs/chapters/04-1.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+\end{bmatrix} = \begin{bmatrix}
+y_1\\
+y_2\\
+y_3\\
+y_4
+\end{bmatrix}
+$$
+
+The above formulation will be difficult to train. Fortunately there are ways to simply. 
+
```
Comment (Atcold) [inline docs/chapters/04-1.md]: One more empty line.

File: docs/chapters/04-1.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+\vdots\\
+x_n
+\end{bmatrix} = \begin{bmatrix}
+y_1\\
+y_2\\
+y_3\\
+y_4
+\end{bmatrix}
+$$
+
```
Comment (Atcold) [inline docs/chapters/04-1.md]: Similarly, add one more line.

File: docs/chapters/04-1.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+## [Listening to Convolutions - Jupyter Notebook](0:33:01-0:59:59)
+
+The Jupyter Notebook can be found [here](https://github.com/Atcold/pytorch-Deep-Learning/blob/master/07-listening_to_kernels.ipynb). Note the names of github repository and conda environment have changed.
+
+In this notebook, we are going to explore Convolution as 'running scalar product'.
+
+Importing library librosa and we can load data $\mathbf{x}$ and sampling rate. In this case, there are 70641 samples, learning rate is 22.05kHz and total time is 3.2s. The imported audio signal is wavy (refer to Fig1) and we can guess what it sounds like from the amplitude of y axis. Listening to x(t), the audio signal is actually the sound played when turning off the Windows system (refer to Fig2). 
+![Fig1 Audio signal](https://i.imgur.com/LLdvx8d.png)
```
Comment (Atcold) [inline docs/chapters/04-1.md]: Images have to be submitted too!
Also, use `<center><br><b>Fig</b> Blah, blah</center>` format.

File: docs/chapters/04-1.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+To get notes from wave from, Fourier transform (FT) is a good guess, however, if FT is played on the whole signal, notes would come out together, the is hard to figure out the exact time and location of each pitch. Localized FT is needed, which is called spectroscope. As is shown in the spectrogram (refer to Fig3), different pitched peak at different values (e.g. first pitch peaks at 1600). Concatenate the four pitches based on their frequencies, we get the reconstruction. 
+![Fig3 Audio sugnal and its spectrum](https://i.imgur.com/i5oWuZH.png)
+
+Convolution of signal with different pitches can help to extract all notes from a given piece, auido matches specific kernels. The spectrograms of origin signal and the conctatenation (refer to Fig4) is shown, as well as the frequencies of real signal and four pitches (refer to Fig5). The plot and sound of convolution of four kernels indicate convolution performance is pretty good (refer to Fig6). 
+
+![Fig4 Spectrogram of real signal and concatenation](https://i.imgur.com/YTyVcH7.png)
+![Fig5 First melody's note](https://i.imgur.com/O4narOi.png)
+![Fig6 COnvolution of four kernels](https://i.imgur.com/R6iJh9N.png)
```
Comment (Atcold) [inline docs/chapters/04-1.md]: All images should be centered and have a caption and be loaded from the repo, not from online, somewhere.

File: docs/chapters/04.md [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+## Lecture part A
+
+No lecture this week.
+
+
+## Lecture part B
+
+No lecture this week.
+
+
```
Comment (Atcold) [inline docs/chapters/04.md]: You can remove this.

**Comment by wwengm** [comment]
Correction applied. Hope the format is up to standard this time.
My apologies that I was unable to test on my local machine so I can not verify the result, but I do have made all the requested corrections. 
Feel free to let me know if any other corrections are needed.

**Comment by Atcold** [comment]
Can you please ask editor 1 to add the
```
                 |
----a^(j)----   a_i
                 |
```
notation?

**Comment by Atcold** [comment]
He can recycle this from [last year](https://github.com/Atcold/SP19-DL-collaborative-notes).

Screenshot 2020-02-26 11 41 48

**Comment by wwengm** [comment]
Sure will do.

**Comment by wwengm** [comment]
Hi Alfredo,

I have trouble converting the vertical and horizontal bars from latex to
markdown. Any suggestions?

I have tried the below code but does not seem to render well in HackMD,
they don't seem to recognize \rule and \vrule command.

$$\vect{z} = {\matr{A}}\vect{x} =
\begin{bmatrix}
    \rule[0.5mm]{0.8cm}{0.1mm} \; {\vect{a}}^{(1)} \;
\rule[0.5mm]{0.8cm}{0.1mm} \\
    \rule[0.5mm]{0.8cm}{0.1mm} \; {\vect{a}}^{(2)} \;
\rule[0.5mm]{0.8cm}{0.1mm} \\
    \vdots \\
    \rule[0.5mm]{0.8cm}{0.1mm} \; {\vect{a}}^{(m)} \;
\rule[0.5mm]{0.8cm}{0.1mm} \\
\end{bmatrix}
\begin{pmatrix}
    \vrule height 0.6cm \\ \vect{x} \\ \vrule height 0.6cm
\end{pmatrix} =
\begin{pmatrix}
    {\vect{a}}^{(1)} \vect{x} \\ {\vect{a}}^{(2)} \vect{x} \\ \vdots
\\ {\vect{a}}^{(m)} \vect{x}
\end{pmatrix}_{m \times 1}$$

**Comment by Atcold** [comment]
Here you have the supported function https://katex.org/docs/supported.html

**Pull request merged**